### DISCUSSION TITLE
Mapping pedal cadence to percent

### DISCUSSION THREAD

**Discussion by EigenMania** [danielnilsson9/bbs-fw#119] [Q&A] [answered]
I'm trying to use the MAP16 definition in util.h to map the pedal cadence to a percentage value: 

`uint8_t pas_cadence_pct = (uint8_t)MAP16(pas_get_cadence_rpm_x10(), 0, MAX_CADENCE_RPM_X10, 0, 100);`
`eventlog_write_data(EVT_DATA_PEDAL_CADENCE_RPM, pas_get_cadence_rpm_x10());`
`eventlog_write_data(EVT_DATA_PEDAL_CADENCE_RPM, pas_cadence_pct);` 

From viewing the event log, it would appear that sometimes the values are mapped correctly (MAX_CADENCE_RPM_X10 = 1680):

![image](https://user-images.githubusercontent.com/23363130/235781371-41eaf2ee-002d-45c5-9231-819e33cbbe7e.png)

But other times the mapped value is just completely wrong (greater than 100):

![image](https://user-images.githubusercontent.com/23363130/235781561-7fab252e-ebd5-499c-8790-5dc69d04253d.png)

Or if the cadence gets high enough the mapped value wraps back to 1: 

![image](https://user-images.githubusercontent.com/23363130/235781746-3a1f39ed-e82a-428c-ab4a-ce80f7eecf37.png)

Am I just mis-using the definition, or making a really dumb mistake in C?

**Comment by danielnilsson9** [comment, marked as answer]
Probably overflow, MAP16 is implemented using int16_t for calculations.

`#define MAP16(x, in_min, in_max, out_min, out_max)	((((int16_t)x) - (in_min)) * ((out_max) - (out_min)) / ((in_max) - (in_min)) + (out_min))`

(x - in_min) * (out_max - out_min) must not be more than 32768 but in your case it will reach 168000. You will have to use MAP32.

MAP16 is mostly useful if you only have 8 bit numbers. When used one must always check that it wont overflow.

**Comment by dav0000000** [comment]
You need MAP32. Something similar is one of the best changes I've made... pas boost. Instead of mapping the rpm to current directly, it just boosts the set pas level current in its own function apply_pas_boost(). 3 variables needed rpm min, rpm max and boost percent. Best settings for me are rpm 52 to 92 with 35% boost of 30amps. Also had to change the cadence calculation in sensors to 1 full revolution otherwise it surged due to rpm speed changing during one pedal revolution and way too many target_current updates. When I pedal my motor is nearly always going way faster than I am actually pedalling... clown pedalling. I haven't tried it with cadence limiting as I let the motor hit max rpm and have that turned off. I'd have to look a bit closer to see if I did anything else. Happy to share code snippets if that interests you.

**Comment by EigenMania** [comment]
Ahh perfect thank you both for explaining and clarifying, it was certainly overflow and MAP32 worked properly. 

The idea I am toying with is to map pedal cadence to the motor commanded speed and tune it in such a way that the motor RPM setpoint either matches the pedalling cadence or just slightly leads it. This should in theory allow the bike to support the rider at any cadence and get rid of ghost pedalling. @dav0000000  for sure, share your branch with me and I'll definitely look at your changes. Always interested in what other ideas are out there.

**Reply by dav0000000** [reply]
Yes, your use case is different as more than 95% of my riding is clown pedalling. I must be honest, I was drawn to this firmware to customise it so that it can't be operated by anyone else, especially 'you know who' so I hope you respect I don't want it public in any way. You probably won't need to smooth the cadence across a revolution if you are not clown pedalling and the part you are doing is not too complicated I think. Good luck.

**Reply by dav0000000** [reply]
Just a side note on this. Will you be relying on cadence speed control in the nec mcu to reduce the power while your code is increasing current as you pedal faster? You might find it will result in no power when you shift to a larger gear. Like preparing for a big hill... and it sucks. I had this problem and turned off cadence speed control a long time ago. I never turned it back on until this morning after reading your comment. I'll be turning it off again. Call me a clown but I easily maintain the same speed in the same assist level across different terrain just by adjusting the pedal rpm and shifting up or down a gear. It's definitely not for everyone this clown pedalling.

**Reply by EigenMania** [reply]
So far I am using the standard functions to compute the current setpoint (apply_pas_cadence() and apply_current_ramp_up()). My preliminary tests have gone quite well, I get power whenever I accelerate the cranks / pedals, but the bike does not out-ride me anymore (which I find especially annoying about cadence-based PAS control at low speeds). But there is still some testing to do. It is a tiny bit jerky, so perhaps a filter on the commanded speed might improve smoothness. Thank you for your tips, I will test as well with shifting and starts/stops on steep hills.
